CONTRIBUTING.md

```diff
@@ -19,17 +36,223 @@ Follow either of the two links above to access the appropriate CLA and
+of the snippet.
+
+Example:
+
+```ruby
+# project_id  = "Your Google Cloud project ID"
+# bucket_name = "Your Google Cloud Storage bucket name"
+  
+require "google/cloud/storage"
+  
+storage = Google::Cloud::Storage.new project: project_id
+```
+
+#### require statements
+
+Reference snippets should include the `require` statement(s) necessary
+to import relevant dependencies for the code sample.
+
+If a snippet uses a Google Cloud client library, it should be required
+in the snippet.  This documents how to require the client library.
+
+Example:
+
+```ruby
+# project_id  = "Your Google Cloud project ID"
+# bucket_name = "Your Google Cloud Storage bucket name"
+
+require "google/cloud/storage"
+
+storage = Google::Cloud::Storage.new project: project_id
+```
+
+**Note:** common libraries such as "date" and "time" do not need to be
+required within the snippet.  Explicitly require dependencies that
+add value and should be documentated.
+
+### Executing snippets
+
+When possible, reference code snippets should be executable.
+
+Each reference code snippet should be executable via a
+corresponding command-line application.
+
+For example, the "create_bucket" snippet in "buckets.rb" should be
+executable by running `bundle exec ruby buckets.rb create`
+
+#### Argument parsing
+
+To make snippets executable, add a postable to the end of the file
+that calls the snippet's method.
+
+For example:
+
+```ruby
+def create_bucket project_id:, bucket_name:
+  # ...
+end
+
+if __FILE__ == $PROGRAM_NAME
+  command = ARGV.shift
+  
+  case command
+  when "create"
+    create_bucket project_id:  ENV["GOOGLE_CLOUD_PROJECT"],
+                  bucket_name: ARGV.first
+  else
+   puts <<-usage
+Usage: bundle exec ruby buckets.rb [command] [arguments]
+
+Commands:
+  create <bucket>    Create a new bucket with the provided name
+  
+Environment variables:
+  GOOGLE_CLOUD_PROJECT must be set to your Google Cloud project ID
+   usage
+  end
+end
+```
+
+Argument parsing is intentionally simple.
+
+### Tutorial applications
+
+Tutorial applications are sample applications that demonstrate how to
+use one or more Google APIs and products to create a fully-functioning
+application.
+
+A tutorial sample may include command-line application(s) or
+web application(s).
+
+The purpose of tutorial applications is very different from reference snippets.
+
+A tutorial application should be an idiomatic Ruby application.
+Code snippets from tutorial applications demonstrate just a *part* of the
+*whole* application are are **not isolated**.
+
+ 1. Tutorial snippet may not include `require` statements
+ 1. Tutorial snippet may not include client library instantiation
+ 1. Tutorial snippet may not include placeholder variables
+
+Here is an small example of the snippets that might be extracted from
+a fully implemented tutorial application:
+
+#### Pub/Sub sample application
+
+##### Create client
+
+```ruby
+require "google/cloud/pubsub"
```

Comment:
**Question**: When would a ruby application **require** and then initialize an _instance_ variable within a few lines?